### ISSUE TITLE
Identify and speed up some of the most intense tests

### ISSUE THREAD

**Issue by dukebody** [dask/dask#1317]
Some tests of the dask test suite are quite computational intensive. If one is running tests on a single core, running them all might get quite time consuming and slows down development. I'm an impatient man. :)

We could spend some effort identifying the most time consuming tests and try to come up with some way to speed them up without sacrificing functionality coverage.

This task is low priority IMO, but might be an easy win.

**Comment by mrocklin** [comment]
Are there tools for pytest that rank tests by duration?

**Comment by dukebody** [comment]
There is an option `  --durations=N         show N slowest setup/test durations (N=0 for all).` I believe we can use.

However it does not measure fixtures. See https://www.mail-archive.com/pytest-dev@python.org/msg01379.html

**Comment by dukebody** [comment]
Output of `py.test dask --duration=20` command:

```
===================================== slowest 20 test durations ======================================
23.77s call     dask/dataframe/tests/test_arithmetics_reduction.py::test_arithmetics_different_index
22.22s call     dask/dataframe/tests/test_arithmetics_reduction.py::test_arithmetics
18.31s call     dask/dataframe/tests/test_dataframe.py::test_cov_corr_stable
11.71s call     dask/dataframe/tests/test_dataframe.py::test_repartition
9.27s call     dask/dataframe/tests/test_groupby.py::test_split_apply_combine_on_series
8.03s call     dask/dataframe/tests/test_arithmetics_reduction.py::test_frame_series_arithmetic_methods
6.09s call     dask/dataframe/tests/test_multi.py::test_merge_by_index_patterns[outer]
6.06s call     dask/dataframe/tests/test_multi.py::test_merge_by_index_patterns[right]
6.04s call     dask/dataframe/tests/test_multi.py::test_merge_by_index_patterns[inner]
6.01s call     dask/dataframe/tests/test_multi.py::test_merge_by_index_patterns[left]
3.13s call     dask/dataframe/tests/test_dataframe.py::test_concat5
3.03s call     dask/dataframe/tests/test_multi.py::test_merge_by_multiple_columns[outer]
3.03s call     dask/bag/tests/test_bag.py::test_groupby_tasks
3.00s call     dask/dataframe/tests/test_multi.py::test_merge_by_multiple_columns[left]
2.99s call     dask/dataframe/tests/test_multi.py::test_merge_by_multiple_columns[inner]
2.96s call     dask/dataframe/tests/test_multi.py::test_merge_by_multiple_columns[right]
2.49s call     dask/dataframe/tests/test_multi.py::test_join_by_index_patterns[outer]
2.49s call     dask/dataframe/tests/test_multi.py::test_join_by_index_patterns[right]
2.47s call     dask/dataframe/tests/test_multi.py::test_join_by_index_patterns[inner]
2.40s call     dask/dataframe/tests/test_multi.py::test_join_by_index_patterns[left]
```

I believe we could try to speed up some of the tests taking 10+ secs.

**Comment by dukebody** [comment]
Most of the time in `dask/dataframe/tests/test_arithmetics_reduction.py` is spend in the functions `check_frame_arithmetics` and `check_series_arithmetics`. 

These functions test if an extensive set of arithmetic operations done with dask dataframes provides same results as with plain pandas dataframes. They take around 1 sec per run (with profiler overhead, without profiler is approx. half that time) and are being run 21 times. I believe we want to keep all arithmetic comparisons, but perhaps we can run each less times, or we can speed up the `eq` test helper function?

Investigating further...

**Comment by dukebody** [comment]
From https://github.com/dask/dask/blob/master/dask/dataframe/utils.py#L227:
```
def eq(a, b, check_names=True, **kwargs):
    assert_divisions(a)  # 33% time of eq function
    assert_divisions(b)
    assert_sane_keynames(a)
    assert_sane_keynames(b)
    a = _check_dask(a, check_names=check_names)  # 47 % time of eq function
    b = _check_dask(b, check_names=check_names)
    if isinstance(a, pd.DataFrame):
        a = _maybe_sort(a)
        b = _maybe_sort(b)
        tm.assert_frame_equal(a, b, **kwargs)
    elif isinstance(a, pd.Series):
        a = _maybe_sort(a)
        b = _maybe_sort(b)
        tm.assert_series_equal(a, b, check_names=check_names, **kwargs)
    elif isinstance(a, pd.Index):
        tm.assert_index_equal(a, b, **kwargs)
    else:
        if a == b:
            return True
        else:
            if np.isnan(a):
                assert np.isnan(b)
            else:
                assert np.allclose(a, b)
return True
```

I checked `assert_divisions` and `_check_dask` functions and apparently both get keys from the dask using a syncrhornous getter:

```
def assert_divisions(ddf):
    ...
    results = get_sync(ddf.dask, ddf._keys())
    ...
```

```
def _check_dask(dsk, check_names=True):
    import dask.dataframe as dd
    if hasattr(dsk, 'dask'):
        result = dsk.compute(get=get_sync)
```
Questions:

 1. Do we really need to use the sync getter here? It is obviously slower than the async one. Furthermore, the async getter is the default, so the tests are not really testing the default case but an easier one, which can lead to unadverted bugs.
 2. We probably don't need to get the dask keys twice. If this is true, we can probably refactor the test helper functions to get it only once and cut the testing time of many functions by half.

Do you think this is worth to spend time on?

**Comment by mrocklin** [comment]
get_sync is nice for two reasons

1.  We can debug, which is critical when tests fail
2.  It has less overhead and so is often faster for small datasets

**Comment by mrocklin** [comment]
Another option here would be to mark some of the slower tests as `slow` so that they always run on travis.ci but don't run by default when calling `py.test` locally without the `--run-slow` flag

**Comment by dukebody** [comment]
OK let's do that for now. I'm marking the top 3 slowest tests as slow.

Another thing that would bring speedups is to break the slow big tests into smaller ones. That way the parallel execution of tests can become faster, since each testing worker gets the same amount of tests. Most of the workers finish very fast with their work and only the ones with slow tests remain, not utilizing all the cpu cores in an optimal fashin.

**Comment by nirizr** [comment]
I think this can be closed now.

**Comment by GTimothee** [comment]
As this issue is still open, can I take my chance of optimizing some of these functions ?

**Comment by TomAugspurger** [comment]
https://github.com/dask/dask/pull/1334 got a few. We'll always improvements to speed up the test suite that doesn't sacrifice correctness.
